place the tm over excview

mmerickel opened the pull request:
The goal here is to fix #40 and hopefully before it is done, #41.

This PR will be pyramid_tm 2.0 and will require pyramid 1.7+. The current implementation has exposed a couple bugs in pyramid 1.7 which I've had to fix unfortunately.

This feature depends on `request.invoke_exception_view` introduced in pyramid 1.7 in order to render exception responses AFTER excview has executed. The workflow here is to add an excview (with low priority) that handles any exception and checks for retryability. It then sets state and propagates the response up to the tm which catches and tries again. If attempts is exhausted then the error is rendered using `request.invoke_exception_view`.

This maintains basically all the guarantees of the original pyramid_tm as far as the responses returned with the added benefit that the transaction is still alive during the excview tween.

I've also added a shim such that if the tm is explicitly added under excview like it is now, thinks should sitll work like they did before.

Currently there is no attempt to copy requests between attempts. I'm worried about performance there because a naive approach would always copy the request. I don't know if there's a way around that except when `tm.attempts == 1`. Input would be greatly appreciated.

bertjwregeer commented:
In the case of a `GET` request, a copy is incredibly cheap since it's a shallow copy, since it's just the `environ` and there is no `wsgi.input`.

In the case of `POST`/`PUT`/`PATCH` with body, `copy()` will make a copy of the `wsgi.input` in turn also making it seekable: http://docs.webob.org/en/latest/api/request.html#webob.request.BaseRequest.copy and (https://github.com/Pylons/webob/blob/master/webob/request.py#L869)

WebOb internally as soon as you touch `request.POST` (or `request.params`) requires that the input is seekable, so if the view requires `.POST` or `.params` the file input would be made seekable anyway (possibly by copying it to a BytesIO or file).

After that any copy will copy the `BytesIO` or the seekable file.

Naively calling `request.make_body_seekable()` shouldn't cause much overhead unless no view matches and the body of the request would never get accessed in the first place...

mmerickel commented:
Are you referring to the current behavior? To fix #41 we need to create a new request object. This is going so far as:

```python
def tm_tween(request):
    orig_request = request
    factory = registry.queryUtility(IRequestFactory, Request)
    for attempt in attempts:
        request = factory(orig_request.environ)
        return handler(request)
```

Is this bad to just use the environ again? Is `request.copy()` preferable? I wanted to use the factory incase a user had supplied a more complex factory via `config.set_request_factory`.

bertjwregeer commented:
You would need to duplicate the logic of `request.copy()` to make that work. Just passing in the `environ` may mean that `environ['wsgi.input']` is in the wrong state (i.e. it has been seeked all the way to the end, or it was never made seekable in the first place.) You also want a copy of the environ, because the application may make changes to it that could alter things further down the line, unless you want to explicitly allow that, in which case you need to take extra care with `wsgi.input`.

If it was never made seekable, or is seeked to the end, the second attempt will not have the request body available, so request.POST or accessing request.body might fail or and do the wrong thing.

As long as the factory returns an object that has `__class__` set to itself, `request.copy()` will do the right thing, or you can replicate the behaviour:

```
def tm_tween(request):
    request.make_body_seekable()
    orig_env = request.environ
    factory = registry.queryUtility(IRequestFactory, Request)
    for attempt in attempts:
        # Copy the environment so that we don't allow downstream tweens to affect the original environment
        request = factory(orig_env.copy())
        # Create a copy of the body, so that any modifications to wsgi.input don't affect the next attempt
        request.copy_body()
        return handler(request)
```

mmerickel commented:
These are my personal expectations. Please correct me if you disagree:

- I don't expect apps to change the body, so sharing it across attempts is something I'm open to doing.
- I don't expect anything in the environ to be isolated to a single attempt.
- I think the request factory should be invoked if we are making a copy if it's possible.

bertjwregeer commented:
1. I would agree, as long as it is made seekable and you `.seek(0)` it, things should be fine. Just call `make_body_seekable()` otherwise if the first request consumes it (by using `request.body`) then all future attempts will get an empty body. I was making the case that if you want a completely clean retry attempt, you'd need to make sure that the body is unique. As long as it is documented I don't think it should be an issue.
2. I'd be okay with that, in my apps I don't make changes to the request environ. It just feels like it becomes a way for something to be passed between attempts, not sure if that is desirable or not, especially since we are already going so far as to rebuild the request object from scratch. I have no dog in this fight.
3. In my example snippet above the request factory is used, and I am onboard with that.

mmerickel commented:
The ideal situation seems to me if we had a structure like the following:

middleware ingress -> tm manager -> pyramid ingress -> excview -> main

Basically for each attempt the request bubbles out of pyramid and back into middleware. The issue that on the final attempt we want to render a pretty error using exception views. Pylons had a crazy middleware that would call back into the app to render the error after the error propagated out. A way around that is if the TM injected into the environ a bit of info about whether this is the last attempt (`environ['tm.last_attempt'] = True`). If so, we can catch it in an excview and render a response instead of bubbling the exception out of the app. This avoids any issues with swapping out the request object after pyramid ingress (which just doesn't feel so great but is currently what we're discussing).

The excview logic is basically exactly what's in this PR, except with the predicate `is_tm_retryable` returning `False` if we are on the last attempt to avoid converting it to a `RetryableError` and instead allowing an exception view to handle it and render a response. The TM would then return the response but abort the transaction.

Interestingly this design does not require the new `invoke_exception_view` api. Unfortunately to accomplish this you need support from middleware which is out of scope of what is possible using `config.include`. We could consider `config.add_middleware(app_filter)` but I don't like doing that for several reasons. The old `repoze.tm2` was middleware, however this is worse because we need to add middleware **and** app-level config so you'd need to add a pipeline to your ini and setup `pyramid.includes`.

Where can we find a middleground? Is swapping the request out inside a tween the best we can do?

bertjwregeer commented:
You are suggesting that the WSGI server itself should retry, if the error is retryable rather than trying to handle that within a tween within Pyramid?

mmerickel commented:
No, a middleware component outside of pyramid would do the retry. This is how repoze.tm2 worked (the code that pyramid_tm replaced). Pretty much the only downside to that approach was 1) the middleware can't see exceptions in pyramid that are squashed by exception views and 2) you can't use your pyramid app to render errors on retryable exceptions, so you have to put an error middleware on top of the tm middleware and now you are just building everything in middleware.

miohtama commented:
Is it possible to expose the current attempt number to the upstream, so that application knows how many times the current request has been replayed? This data may be required for testing (functional test your application behaves correctly in the case of replay) and diagnostics (seeing that certain view is causing high number of replays)?

mmerickel commented:
@miohtama I had thought about and planned to add a feature in which the underlying code could tell if it was the last attempt. I hadn't planned to add the exact attempt number.

mmerickel commented:
I have reworked this PR to be almost feature complete. It now makes a copy of the request on subsequent attempts, handles threadlocals pointing to the new request and some other quirks of making a new request object.

I've added a new predicate `tm_last_attempt=True/False` which user-code can use to register exceptions only for one case or the other if they wish. I plan to add another as well that would be `tm_exc_is_retryable=True/False` which will be capable of telling you if the exception is retryable if for some reason you want to know. This would allow you to add an exception view for something like `context=Exception, tm_exc_is_retryable=True` to render an appropriate error response for a failed retryable exception which would suggest to the user to try again.

Also the `request.environ` contains 3 keys: `tm.active`, `tm.attempt` and `tm.attempts` which can be used to determine where you're at in the cycle, as requested by @miohtama.

I would greatly appreciate some review of this as it will likely affect your applications.

miohtama commented:
Cool stuff. I have written a py.test + SQLAlchemy thread based conflcit test suite for transaction retries here:

https://github.com/websauna/websauna/blob/master/websauna/tests/test_retry.py

I could try to port these tests to exercise WebTest + Waitress + pyramid_tm, as I feel most pyramid_tm users are interested in SQLAlchemy support. However I do not know if it's desirable to have these in pyramid_tm, zope.sqlalchemy or somewhere else.

mmerickel commented:
@miohtama I would happily merge a PR that added some of those tests to the pyramid_tm test suite. We could depend on zope.sqlalchemy, etc in the testing extras.

mmerickel commented:
I've cleaned up the implementation. I'm planning to release this as 1.0a1.

mmerickel commented:
I released 1.0 from current master so people have something to pin to if they want. This will be 2.0a1 instead.

mmerickel commented:
The main thing this doesn't currently solve is finished/response callbacks. I think I need to manually execute response/finished callbacks when we finish an intermediate attempt. For the final attempt I need to copy the request state back to the original request so that it is executed as normal. This is crazy, but poking at these private apis is the only way to make things happen until pyramid's request pipeline becomes more flexible (moving finished and response callbacks to tweens so that pyramid_tm can cover them).

bertjwregeer commented on .travis.yml:
```diff
@@ -15,16 +15,6 @@ matrix:
-        - python: 2.7
-          env: TOXENV=py27-pyramid15
-        - python: 3.5
-          env: TOXENV=py35-pyramid16
```
Is this a sign of things to come with compatibility requiring Pyramid 1.7 or higher?

mmerickel replied:
https://github.com/mmerickel/pyramid_tm/blob/feature/isolate-requests/pyramid_tm/__init__.py#L96-L115

bertjwregeer commented on pyramid_tm/__init__.py:
```diff
@@ -1,15 +1,21 @@
+from pyramid.request import Request, apply_request_extensions
 from pyramid.settings import asbool
-from pyramid.util import DottedNameResolver
+from pyramid.threadlocal import manager as request_manager
```
I know `pyramid_tm` is special, but this is not public API, by using it here are we not de facto making it a public API?

mmerickel replied:
Yes, placing the TM above exception views leverages `request.invoke_exception_view` to render exceptions that may occur after `EXCVIEW` tween but before pyramid_tm.

mmerickel replied:
It's not unprecedented for official pyramid addons to use private pyramid apis, despite being ugly and unfortunate... it's necessary here until pyramid grows some other way to copy the request.

bertjwregeer commented on pyramid_tm/__init__.py:
```diff
@@ -52,67 +64,221 @@ def tm_tween(request):
+
+        # hang onto a reference to the original request as it's the thing
+        # used above the tm tween, we can't change that
+        orig_request = request
```
This feels so incredibly wrong...

bertjwregeer commented on pyramid_tm/__init__.py:
```diff
@@ -52,67 +64,221 @@ def tm_tween(request):
+            # in the request factory and elsewhere if people want it
+            # note: set all of these values here as they are cleared after
+            # each attempt
+            environ = request.environ
```
Since we blow away `request`, could this not be `orig_request.environ` instead? Setting it on `orig_request.environ['xxx']` would possibly be useful as well so that things above `pyramid_tm` can still get at the `tm` information stored in the `environ` (in case they are aware of them).

Maybe not worth it?

bertjwregeer commented on pyramid_tm/__init__.py:
```diff
@@ -52,67 +64,221 @@ def tm_tween(request):
+            # avoid extra copies of the body
+            if number > 0:
+                # try to make sure this code stays in sync with pyramid's
+                # router which normally creates requests
```
Oh man...

bertjwregeer commented on pyramid_tm/__init__.py:
```diff
@@ -52,67 +64,221 @@ def tm_tween(request):
+
+                # check for a squashed exception and handle it
+                # this would happen if an exception view was invoked and
+                # rendered an error response
```
So if the user has their own exception view, they would need to be aware of the tm to make sure they don't do anything unless it's a last attempt.

mmerickel replied:
Only if the logic in their exception view can't be rolled back. If that's the case then they can use the predicates as well as `pyramid_tm.exc_is_retryable`, or check the environ values to see if it's the last attempt. Otherwise exception views can just execute like normal.

bertjwregeer commented on pyramid_tm/__init__.py:
```diff
@@ -52,67 +64,221 @@ def tm_tween(request):
+                        return render_exception(request, exc_info)
+
                 finally:
+                    # keep the manager alive until after invoke_exception_view
```
s/invoke_exception_view/render_exception/

?

bertjwregeer commented on pyramid_tm/__init__.py:
```diff
@@ -52,67 +64,221 @@ def tm_tween(request):
+
+                # propagate exception info back to the original request,
+                # possibly clearing out a retryable error triggered from the
+                # first attempt
```
:-( I understand why, but this seems like it can be error prone... how about any finished callbacks and all that stuff? Those are set on this latest request, not on the original request...

You are right when you said the callbacks need to be under pyramid_tm.

bertjwregeer commented on pyramid_tm/__init__.py:
```diff
@@ -52,67 +64,221 @@ def tm_tween(request):
+    return environ['tm.attempt'] == environ['tm.attempts'] - 1
+
+
+class RetryableExceptionPredicate(object):
```
I like this, better than needing to add boilerplate to my views

bertjwregeer commented on pyramid_tm/__init__.py:
```diff
@@ -52,67 +64,221 @@ def tm_tween(request):
+        )
+
+
+class LastAttemptPredicate(object):
```
Excellent!

bertjwregeer commented on setup.py:
```diff
@@ -24,7 +24,7 @@
 
 install_requires = [
-    'pyramid>=1.2dev',
+    'pyramid>=1.7b4',
```
That answers my question about the .travis changes above...

mmerickel commented:
I've replaced this with pyramid_retry and #55.
